File: macos/Sources/LumaMDCore/MathRenderer.swift
import Foundation

internal enum MathRenderer {
    static func renderInline(_ expression: String) -> String {
        "<span class=\"math-inline\" role=\"math\" aria-label=\"\(escape(label(for: expression)))\">\(math(for: expression))</span>"
    }

    static func renderBlock(_ expression: String) -> String {
        "<div class=\"math-block\" role=\"math\" aria-label=\"\(escape(label(for: expression)))\">\(math(for: expression))</div>"
    }

    private static func math(for expression: String) -> String {
        var parser = MathMLParser(expression)
        return "<math xmlns=\"http://www.w3.org/1998/Math/MathML\">\(parser.render())</math>"
    }

    private static func label(for expression: String) -> String {
        let replacements = [
            ("\\frac", "fraction "), ("\\sqrt", "square root "),
            ("\\sum", "sum "), ("\\prod", "product "), ("\\int", "integral "),
            ("\\mathrm", ""), ("\\mathbf", ""), ("\\operatorname", ""),
            ("{", "("), ("}", ")"),
        ]
        var value = expression.unicodeScalars.filter { $0.value >= 0x20 && $0.value != 0x7F }
            .map(String.init).joined()
        for (source, replacement) in replacements {
            value = value.replacingOccurrences(of: source, with: replacement)
        }
        let compact = value.replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
        return compact.isEmpty ? "formula" : compact
    }

    private static func escape(_ value: String) -> String {
        InlineMarkdown.escape(value)
    }
}

File: app/src/main/java/dev/lumamd/viewer/core/MathRenderer.java
package dev.lumamd.viewer.core;

final class MathRenderer {
    private MathRenderer() {
    }

    static String renderInline(String expression) {
        return "<span class=\"math-inline\" role=\"math\" aria-label=\""
                + InlineMarkdown.escape(label(expression)) + "\">"
                + math(expression) + "</span>";
    }

    static String renderBlock(String expression) {
        return "<div class=\"math-block\" role=\"math\" aria-label=\""
                + InlineMarkdown.escape(label(expression)) + "\">"
                + math(expression) + "</div>";
    }

    private static String math(String expression) {
        return "<math xmlns=\"http://www.w3.org/1998/Math/MathML\">"
                + new MathMLParser(expression).render()
                + "</math>";
    }

    private static String label(String expression) {
        String value = expression
                .replace("\\frac", "fraction ")
                .replace("\\tfrac", "fraction ")
                .replace("\\sqrt", "square root ")
                .replace("\\sum", "sum ")
                .replace("\\prod", "product ")
                .replace("\\int", "integral ")
                .replace("\\mathrm", "")
                .replace("\\mathbf", "")
                .replace("\\mathcal", "")
                .replace("\\operatorname", "")
                .replace('{', '(')
                .replace('}', ')');
        StringBuilder printable = new StringBuilder(value.length());
        for (int index = 0; index < value.length(); index++) {
            char character = value.charAt(index);
            if (character >= 0x20 && character != 0x7F) {
                printable.append(character);
            }
        }
        String compact = printable.toString().replaceAll("\\s+", " ").trim();
        return compact.isEmpty() ? "formula" : compact;
    }
}
